ddl/column_with_index.go

```diff
@@ -0,0 +1,683 @@
+					newIndexes = append(newIndexes, idx)
+				}
+			}
+			tblInfo.Indices = newIndexes
+
+			for _, idx := range ctidxInfos {
+				// Rename temp index
+				idx.Name = renameTempCompositeIdxToOrigin(idx, cidxInfos)
+				// Set state to public
+				idx.State = model.StatePublic
+			}
+			indexIDs = append(indexIDs, indexInfosToIDList(cidxInfos)...)
+		}
+
+		// All reorganization jobs are done, drop this column.
+		tblInfo.MoveColumnInfo(colInfo.Offset, len(tblInfo.Columns)-1)
+		tblInfo.Columns = tblInfo.Columns[:len(tblInfo.Columns)-1]
+		colInfo.State = model.StateNone
+
+		ver, err = updateVersionAndTableInfo(d, t, job, tblInfo, colOriginalState != colInfo.State)
+		if err != nil {
+			return ver, errors.Trace(err)
+		}
+
+		// Finish this job.
+		if job.IsRollingback() {
+			job.FinishTableJob(model.JobStateRollbackDone, model.StateNone, ver, tblInfo)
+		} else {
+			// We should set related index IDs for job
+			job.FinishTableJob(model.JobStateDone, model.StateNone, ver, tblInfo)
+			job.Args = append(job.Args, indexIDs, getPartitionIDs(tblInfo))
+		}
+	default:
+		return ver, errors.Trace(dbterror.ErrInvalidDDLJob.GenWithStackByArgs("table", tblInfo.State))
+	}
+	job.SchemaState = colInfo.State
+	return ver, errors.Trace(err)
+}
+
+// getCurrentReorgIndex returns the current reorging index, if current reorging index not in
+// jobs reorg index array, it will treat it as all index is reorged.
+func getCurrentReorgIndex(job *model.Job, idxInfos []*model.IndexInfo) (*model.IndexInfo, bool) {
+	reorgIdxIDs := job.GetNeedReorgIndexIDs()
+	pos := job.CurrentReorgIndexIdx
+	if pos >= len(reorgIdxIDs) {
+		// Nothing need to reorg just return all done
+		return nil, true
+	}
+	idxID := reorgIdxIDs[pos]
+	for _, idx := range idxInfos {
+		if idx.ID == idxID {
+			return idx, false
+		}
+	}
+	// Not found the index info need reorg just return all done
+	// Actually should not hit here.
+	return nil, true
+}
+
+// updateReorgDone is update the need reorg index array position and check
+// if all indexes is reorged. If not move to next one and reset reorg related data.
+func updateReorgDone(job *model.Job, indexInfo *model.IndexInfo) bool {
+	reorgIdxIDs := job.GetNeedReorgIndexIDs()
+	// Check reorg done index is in job.NeedReorgIndexIDs
+	inList := false
+	for _, idxID := range reorgIdxIDs {
+		if indexInfo.ID == idxID {
+			inList = true
+			break
+		}
+	}
+	if !inList {
+		return false
+	}
+
+	job.CurrentReorgIndexIdx++
+	if job.CurrentReorgIndexIdx < len(reorgIdxIDs) {
+		// Move to next then update reorg related struct
+		newReorgMeta := &model.DDLReorgMeta{
+			SQLMode:       job.ReorgMeta.SQLMode,
+			Warnings:      job.ReorgMeta.Warnings,
+			WarningsCount: job.ReorgMeta.WarningsCount,
+			Location:      job.ReorgMeta.Location,
+			ReorgTp:       job.ReorgMeta.ReorgTp,
+			IsDistReorg:   job.ReorgMeta.IsDistReorg,
+		}
+		// Reset reorg meta and snapshot version
+		job.ReorgMeta = newReorgMeta
+		job.SnapshotVer = 0
```

Comment:
I have a question here. Set SnapshotVer to zero will re-initialize the `reorgInfo`. At `reorgInfo` initialize function, the snapshot version will get from store. So I don't know the SnapshotVer is same as prev index reorg process. And if the SnapshotVer is not same is there has some issue at reorg process? Such as missing some column to create the index?